### PR TITLE
Refactor `EventContext`

### PR DESCRIPTION
**Pull request by erikjohnston**
Refactor how the `EventContext` class works, with the intention of reducing the amount of state we fetch from the DB during event processing.

The idea here is to get rid of the cached `current_state_ids` and `prev_state_ids` that live in the `EventContext`, and instead defer straight to the database (and its caching). 

One change that may have a noticeable effect is that we now no longer prefill the `get_current_state_ids` cache on a state change. However, that query is relatively light, since its just a case of reading a table from the DB (unlike fetching state at an event which is more heavyweight). For deployments with workers this cache isn't even used.

Part of #12684

### CODE REVIEW

File: synapse/events/snapshot.py [hunk 1]
```diff
@@ -198,13 +180,10 @@ def deserialize(storage: "Storage", input: JsonDict) -> "EventContext":
         Returns:
             The event context.
         """
-        context = _AsyncEventContextImpl(
```
Comment (clokep) [inline synapse/events/snapshot.py:196]: So happy to see this logic disappear. 🔥

File: synapse/events/snapshot.py [hunk 2]
```diff
@@ -57,6 +57,9 @@ class EventContext:
             If ``state_group`` is None (ie, the event is an outlier),
             ``state_group_before_event`` will always also be ``None``.
 
+        delta_before_after: If `state_group` and `state_group_before_event` are not None
+            then this is the delta of the state between the two groups.
```
Comment (clokep) [inline synapse/events/snapshot.py:62]: Is there a better name we can use? Maybe `state_delta_at_event` or something?

File: synapse/storage/persist_events.py [hunk 3]
```diff
@@ -829,15 +804,13 @@ async def _get_new_state_after_events(
...
 
         # Now that we have calculated new_state_groups we need to get
         # their state IDs so we can resolve to a single state set.
-        missing_state = new_state_groups - set(state_groups_map)
-        if missing_state:
-            group_to_state = await self.state_store._get_state_for_groups(missing_state)
-            state_groups_map.update(group_to_state)
+        state_groups_map = await self.state_store._get_state_for_groups(
+            new_state_groups
+        )
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/storage/persist_events.py:813]: Yeah exactly. If some of the state is already cached then `_get_state_for_groups` will use them anyway, so no point trying duplicating that logic here

File: synapse/events/snapshot.py [hunk 4]
```diff
@@ -255,8 +234,10 @@ async def get_current_state_ids(self) -> Optional[StateMap[str]]:
         if self.rejected:
             raise RuntimeError("Attempt to access state_ids of rejected event")
 
-        await self._ensure_fetched()
-        return self._current_state_ids
+        if self._state_group is None:
+            return None
+
+        return await self._storage.state.get_state_ids_for_group(self._state_group)
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/events/snapshot.py]: Perhaps, though if the `EventContext` lives long enough for things to fall out of the cache, then err...

File: synapse/events/snapshot.py (hunk 2 above)
Comment (erikjohnston) [reply synapse/events/snapshot.py:62]: It's hard because I want to make sure people don't think that its the state delta since the `prev_group`

Comment (clokep) [reply synapse/events/snapshot.py:62]: Right! Just the `before_after` I find confusing. Before/after *what*?!

Comment (erikjohnston) [reply synapse/events/snapshot.py:62]: Oh i see. Yes. Ermh, `delta_before_and_after_event`?

File: synapse/events/snapshot.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (clokep) [reply synapse/events/snapshot.py]: I think we don't need to worry about it here, just thought I'd mention it!

**Review by clokep — APPROVED** [review]
I think this is OK? We talked about the naming a bit over chat, leaving it up to you.

**Pull request merged**